### PR TITLE
Lay foundations for re-enabling windows tests: Start CSI-Proxy as a background Job.

### LINKED ISSUE #121: fake CSIProxyMounter to enable unit test
**Is your feature request related to a problem?/Why is this needed**

**Describe the solution you'd like in detail**

Need to fake `CSIProxyMounter`, same as [fakeMounter](https://github.com/kubernetes-csi/csi-driver-smb/blob/65f63946ba5998d4160c8668ec413c2cccc77ecc/pkg/smb/fake_mounter.go#L26-L28), and enable a few unit tests on Windows, current unit tests failed with `return fmt.Errorf("could not cast to csi proxy class"`

```
C:\Users\xiazhang\go\src\github.com\kubernetes-csi\csi-driver-smb\pkg\smb>git diff
diff --git a/pkg/smb/nodeserver_test.go b/pkg/smb/nodeserver_test.go
index ecaeab12..f9a9f386 100644
--- a/pkg/smb/nodeserver_test.go
+++ b/pkg/smb/nodeserver_test.go
@@ -22,9 +22,11 @@ import (
        "fmt"
        "os"
        "reflect"
+       "runtime"
        "syscall"
        "testing"

+       "github.com/kubernetes-csi/csi-driver-smb/pkg/mounter"
        "github.com/kubernetes-csi/csi-driver-smb/test/utils/testutil"

        "github.com/container-storage-interface/spec/lib/go/csi"
@@ -198,7 +200,7 @@ func TestNodeExpandVolume(t *testing.T) {
 }

 func TestNodePublishVolume(t *testing.T) {
-       skipIfTestingOnWindows(t)
+       //skipIfTestingOnWindows(t)
        volumeCap := csi.VolumeCapability_AccessMode{Mode: csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER}
        errorMountSource := "./error_mount_source"
        alreadyMountedTarget := "./false_is_likely_exist_target"
@@ -284,9 +286,15 @@ func TestNodePublishVolume(t *testing.T) {
        // Setup
        _ = makeDir(alreadyMountedTarget)
        d := NewFakeDriver()
-       fakeMounter := &fakeMounter{}
-       d.mounter = &mount.SafeFormatAndMount{
-               Interface: fakeMounter,
+       if runtime.GOOS == "windows" {
+               csiProxyMounter, _ := mounter.NewCSIProxyMounter()
+               d.mounter = &mount.SafeFormatAndMount{
+                       Interface: csiProxyMounter,
+               }
+       } else {
+               d.mounter = &mount.SafeFormatAndMount{
+                       Interface: &fakeMounter{},
+               }
        }

        for _, test := range tests {
```

Without fake, there would be panic:
```
C:\Users\xiazhang\go\src\github.com\kubernetes-csi\csi-driver-smb\pkg\smb>go test -run TestNodePublishVolume
--- FAIL: TestNodePublishVolume (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x10 pc=0x8ae83c]

goroutine 22 [running]:
testing.tRunner.func1(0xc000150100)
        c:/go/src/testing/testing.go:874 +0x3aa
panic(0x996b00, 0xeebb70)
        c:/go/src/runtime/panic.go:679 +0x1c0
github.com/kubernetes-csi/csi-driver-smb/pkg/mounter.(*CSIProxyMounter).ExistsPath(0xc000067e10, 0xa51ca5, 0x8, 0x0, 0x38, 0x8)
        C:/Users/xiazhang/go/src/github.com/kubernetes-csi/csi-driver-smb/pkg/mounter/safe_mounter_windows.go:210 +0x17c
github.com/kubernetes-csi/csi-driver-smb/pkg/mounter.(*CSIProxyMounter).IsLikelyNotMountPoint(0xc000067e10, 0xa51ca5, 0x8, 0xc000093420, 0x40cdf0, 0xc00006d640)
        C:/Users/xiazhang/go/src/github.com/kubernetes-csi/csi-driver-smb/pkg/mounter/safe_mounter_windows.go:148 +0xdf
github.com/kubernetes-csi/csi-driver-smb/pkg/smb.(*Driver).ensureMountPoint(0xc000093858, 0xa51ca5, 0x8, 0x1, 0x2, 0xc00006d640)
        C:/Users/xiazhang/go/src/github.com/kubernetes-csi/csi-driver-smb/pkg/smb/nodeserver.go:274 +0x70
github.com/kubernetes-csi/csi-driver-smb/pkg/smb.(*Driver).NodePublishVolume(0xc000093858, 0xb1fe80, 0xc0000720b0, 0xc0000938e8, 0x1, 0xb135c0, 0xc0000c2d70)
        C:/Users/xiazhang/go/src/github.com/kubernetes-csi/csi-driver-smb/pkg/smb/nodeserver.go:71 +0x1f4
github.com/kubernetes-csi/csi-driver-smb/pkg/smb.TestNodePublishVolume(0xc000150100)
        C:/Users/xiazhang/go/src/github.com/kubernetes-csi/csi-driver-smb/pkg/smb/nodeserver_test.go:301 +0x7f1
testing.tRunner(0xc000150100, 0xa7d240)
        c:/go/src/testing/testing.go:909 +0xd0
created by testing.(*T).Run
        c:/go/src/testing/testing.go:960 +0x357
exit status 2
FAIL    github.com/kubernetes-csi/csi-driver-smb/pkg/smb        1.950s
```

Another way is use real CSIProxyMounter, and run csi-proxy.exe on Windows before running unit tests.
follow guide here: 
  - https://github.com/kubernetes-csi/csi-proxy/tree/master#command-line-options
  - https://github.com/kubernetes-csi/csi-proxy/blob/25491d0543e53381aab69ee5e5fa114182787b43/cmd/csi-proxy/main.go#L22-L24

Or run as a service using nssm:
 - https://github.com/Azure/aks-engine/blob/master/parts/k8s/windowscsiproxyfunc.ps1

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by boddumanohar**
**What type of PR is this?**

/kind test

**What this PR does / why we need it**:
Currently we are skipping a lot of test cases in windows. This is because using a fakeMounter to test on windows is causing isues. This PR lays a basic foundations to run windows based tests related to node by running [CSI-Proxy](https://github.com/kubernetes-csi/csi-proxy) as a background job and the all the mount related operation will be handled by this proxy. For more in-depth design details for CSI-Proxy, please refer to the design doc: https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190714-windows-csi-support.md

**Which issue(s) this PR fixes**:

**Requirements**:
- [X] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [X] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

Added a script to `.github.com/workflows/windows.yml` that downloads and starts CSI-Proxy as windows background job.

**Release note**:
```
none
```

### CODE REVIEW

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -18,10 +18,26 @@ jobs:
...
+        uses: actions/checkout@v2
       - name: Build Test
         run: |
           make smb-windows
       - name: Run Windows Unit Tests
         run: |
+          # start the CSI Proxy before running tests on windows
+          # TODO: (boddumanohar) CSI-Proxy team are not building the csi-proxy.exe. Currently I am using a binary that I built. In future we should change to offcial repo.
+          Start-Job -Name CSIProx -ScriptBlock {
+            Invoke-WebRequest https://github.com/boddumanohar/csi-proxy/releases/download/v0.2.1/csi-proxy.zip -OutFile csi-proxy.zip;
```
Comment (andyzhangx) [inline .github/workflows/windows.yml:30]: could you use this binary: `https://kubernetesartifacts.azureedge.net/csi-proxy/v0.2.0/binaries/csi-proxy.tar.gz` ?

File: pkg/smb/nodeserver_test.go [hunk 2]
```diff
@@ -347,6 +393,7 @@ func TestNodeUnpublishVolume(t *testing.T) {
 	for _, test := range tests {
 		_, err := d.NodeUnpublishVolume(context.Background(), &test.req)
 		if !reflect.DeepEqual(err, test.expectedErr) {
+			fmt.Println(err)
```
Comment (andyzhangx) [inline pkg/smb/nodeserver_test.go:396]: remove this

File: pkg/smb/nodeserver_test.go [hunk 3]
```diff
@@ -198,108 +200,152 @@ func TestNodeExpandVolume(t *testing.T) {
...
+		"linux":   "./target_test",
+		"windows": "C:\\var\\lib\\kubelet\\target_test",
+	}
+
+	platform := "linux"
+	if runtime.GOOS == "windows" {
+		platform = "windows"
+	}
+
+	fmt.Println(platform)
```
Comment (andyzhangx) [inline pkg/smb/nodeserver_test.go:231]: remove this

File: pkg/smb/nodeserver_test.go [hunk 4]
```diff
@@ -198,108 +200,152 @@ func TestNodeExpandVolume(t *testing.T) {
...
+		"linux":   "./error_mount_source",
+		"windows": "C:\\var\\lib\\kubelet\\error_mount_source",
+	}
+	alreadyMountedTarget := map[string]string{
+		"linux":   "./false_is_likely_exist_target",
+		"windows": "C:\\var\\lib\\kubelet\\fake_is_likely_exist_target",
+	}
+	smbFile := map[string]string{
+		"linux":   "./smb.go",
+		"windows": "C:\\var\\lib\\kubelet\\smb.go",
```
Comment (andyzhangx) [inline pkg/smb/nodeserver_test.go:215]: why using this path `C:\\var\\lib\\kubelet\\` on windows?

Comment (andyzhangx) [reply pkg/smb/nodeserver_test.go:215]: another way is using GetWorkDirPath: https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/6e64aea79994672fcc356588d1c652db26fb9136/pkg/azurefile/nodeserver_test.go#L87

Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:215]: the reason why I am using that path is because, within the CSI-Proxy logs I get this error: 
```
PS C:\Users\manoh\github\csi-proxy\bin> .\csi-proxy.exe -v 9
I0925 19:31:16.912285   17948 main.go:51] Starting CSI-Proxy Server ...
I0925 19:31:16.913292   17948 main.go:52] Version: v0.2.1-2-g25491d0
I0925 19:31:39.611493   17948 server.go:122] calling PathExists with path "C:\\Users\\manoh\\github\\csi-driver-smb\\pkg\\smb\\smg.go"
E0925 19:31:39.611493   17948 server.go:125] failed validatePathWindows path: C:\Users\manoh\github\csi-driver-smb\pkg\smb\smg.go is not within context path: C:\var\lib\kubelet

```

It explicitly tells here the path is supposed to be `C:\\var\\lib\\kubelet\\`,

Comment (andyzhangx) [reply pkg/smb/nodeserver_test.go:215]: then what about use func like this, that would be more clear in logic:
```
func getTestPath(path string) string {
	if runtime.GOOS == "windows" {
		return filepath.Join("C:\\var\\lib\\kubelet\\", path)
	}
	return filepath.Join("./", path)
}
```

File: pkg/smb/nodeserver_test.go [hunk 5]
```diff
@@ -241,7 +261,8 @@ func TestNodePublishVolume(t *testing.T) {
 				TargetPath:        smbFile,
 				StagingTargetPath: sourceTest,
 				Readonly:          true},
-			expectedErr: status.Errorf(codes.Internal, "Could not mount target \"./smb.go\": mkdir ./smb.go: not a directory"),
+			expectedErrWindows: nil,
```
Comment (andyzhangx) [inline pkg/smb/nodeserver_test.go:264]: why it's nil on windows?

Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:264]: The CSI-Proxy is not returning any error for the above. Here are the CSI-Proxy logs:
```
I0925 20:14:06.058327    3480 main.go:51] Starting CSI-Proxy Server ...
I0925 20:14:06.058327    3480 main.go:52] Version: v0.2.0-0-ga37f58f
I0925 20:14:11.417233    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\smb.go"
I0925 20:14:11.418235    3480 server.go:220] calling IsMountPoint with path "C:\\var\\lib\\kubelet\\smb.go"
I0925 20:14:11.420241    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\smb.go"
I0925 20:14:11.427243    3480 server.go:172] calling Rmdir with path "C:\\var\\lib\\kubelet\\smb.go"
I0925 20:14:11.428252    3480 server.go:193] calling LinkPath with targetPath "C:\\var\\lib\\kubelet\\smb.go" sourcePath "C:\\var\\lib\\kubelet\\source_test"
I0925 20:14:11.429237    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.430255    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.432234    3480 server.go:172] calling Rmdir with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.438235    3480 server.go:193] calling LinkPath with targetPath "C:\\var\\lib\\kubelet\\target_test" sourcePath "C:\\var\\lib\\kubelet\\error_mount_source"
I0925 20:14:11.439234    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.439234    3480 server.go:220] calling IsMountPoint with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.440235    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.441235    3480 server.go:172] calling Rmdir with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.441235    3480 server.go:193] calling LinkPath with targetPath "C:\\var\\lib\\kubelet\\target_test" sourcePath "C:\\var\\lib\\kubelet\\source_test"
I0925 20:14:11.442243    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\false_is_likely_exist_target"
I0925 20:14:11.443232    3480 server.go:220] calling IsMountPoint with path "C:\\var\\lib\\kubelet\\false_is_likely_exist_target"
I0925 20:14:11.443232    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\false_is_likely_exist_target"
I0925 20:14:11.444234    3480 server.go:172] calling Rmdir with path "C:\\var\\lib\\kubelet\\false_is_likely_exist_target"
I0925 20:14:11.444234    3480 server.go:193] calling LinkPath with targetPath "C:\\var\\lib\\kubelet\\false_is_likely_exist_target" sourcePath "C:\\var\\lib\\kubelet\\source_test"
I0925 20:14:11.445234    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.445234    3480 server.go:220] calling IsMountPoint with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.446233    3480 server.go:122] calling PathExists with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.454239    3480 server.go:172] calling Rmdir with path "C:\\var\\lib\\kubelet\\target_test"
I0925 20:14:11.454239    3480 server.go:193] calling LinkPath with targetPath "C:\\var\\lib\\kubelet\\target_test" sourcePath "C:\\var\\lib\\kubelet\\source_test"
```

Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:264]: Let me find out the exact reason for this behavior and will get back to you.

Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:264]: @andyzhangx This is because, on windows during `Mount` operation is called, under the hood we are actually calling linkPath. https://github.com/kubernetes-csi/csi-driver-smb/blob/master/pkg/mounter/safe_mounter_windows.go#L96. Linking a filepath to a directory which will succeed. Where as in unix we are actually performing an mount from file to a folder which will fail. 

Do you have any suggestions on how this can fixed?

Comment (andyzhangx) [reply pkg/smb/nodeserver_test.go:264]: mount behavior between Linux and Windows is a little different, that's ok. Could you write code like this:
```
[]struct {
		desc               string
		req                csi.NodePublishVolumeRequest
		expectedErr   	   error
		expectedErrLinux   error
		expectedErrWindows error
	}
...
expectedErr := test.expectedErr
if test.expectedErrWindows != test.expectedErrLinux {
	if runtime.GOOS == "windows" {
		expectedErr = test.expectedErrWindows
	} else {
		expectedErr = test.expectedErrLinux
	}
}
```

So in most cases, we only need to define `expectedErr`, and only need to define `expectedErrLinux` and `expectedErrWindows` when it's different. That would save test code.

File: pkg/smb/nodeserver_test.go [hunk 6]
```diff
@@ -198,41 +201,58 @@ func TestNodeExpandVolume(t *testing.T) {
...
 	volumeCap := csi.VolumeCapability_AccessMode{Mode: csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER}
-	errorMountSource := "./error_mount_source"
-	alreadyMountedTarget := "./false_is_likely_exist_target"
-	smbFile := "./smb.go"
-	sourceTest := "./source_test"
-	targetTest := "./target_test"
+
+	// For CSI-Proxy(windows), the path has to be absolute and it should be within the context of value passed for
+	// csi-proxy argument: --kubelet-pod-path. In our tests we are using `C:\var\lib\kubelet\`
+	getTestPath := func(path string) string {
```
Comment (andyzhangx) [inline pkg/smb/nodeserver_test.go:208]: pls move this function outside(standalone), other test cases would also needs this function, e.g. `TestNodeUnpublishVolume`

File: pkg/smb/nodeserver_test.go [hunk 7]
```diff
@@ -198,18 +210,20 @@ func TestNodeExpandVolume(t *testing.T) {
...
 
 	tests := []struct {
-		desc        string
-		req         csi.NodePublishVolumeRequest
-		expectedErr error
+		desc               string
+		req                csi.NodePublishVolumeRequest
+		expectedErr        error
+		expectedErrUnix    error
+		expectedErrWindows error
```
Comment (mayankshah1607) [inline pkg/smb/nodeserver_test.go:226]: We currently have code to differentiate Windows and Linux errors. Try taking a look at - https://github.com/kubernetes-csi/csi-driver-smb/blob/master/test/utils/testutil/testutil.go#L28

This object holds the error messages for both Windows and Linux as a single entity. Also try looking at the following helper methods for assertion. 
Why don't we try reusing that instead of re-introducing the same funcitonility?

Here's an example of this - https://github.com/kubernetes-csi/csi-driver-smb/blob/master/pkg/smb/nodeserver_test.go#L60

File: .github/workflows/windows.yml [hunk 8]
```diff
@@ -18,10 +18,25 @@ jobs:
...
-        uses: actions/checkout@v2 
+        uses: actions/checkout@v2
       - name: Build Test
         run: |
           make smb-windows
       - name: Run Windows Unit Tests
         run: |
+          # start the CSI Proxy before running tests on windows
+          Start-Job -Name CSIProx -ScriptBlock {
+            Invoke-WebRequest https://kubernetesartifacts.azureedge.net/csi-proxy/v0.2.0/binaries/csi-proxy.tar.gz -OutFile csi-proxy.tar.gz;
```
Comment (andyzhangx) [inline .github/workflows/windows.yml:29]: use `v0.2.1`

File: pkg/smb/nodeserver_test.go [hunk 9]
```diff
@@ -304,15 +339,16 @@ func TestNodePublishVolume(t *testing.T) {
 }
 
 func TestNodeUnpublishVolume(t *testing.T) {
-	skipIfTestingOnWindows(t)
-	errorTarget := "./error_is_likely_target"
-	targetFile := "./abc.go"
-	targetTest := "./target_test"
+	errorTarget := getTestPath("error_is_likely_target")
+	targetFile := getTestPath("abc.go")
+	targetTest := getTestPath("target_test")
```
Comment (mayankshah1607) [inline pkg/smb/nodeserver_test.go:344]: We already have utility for generating test paths - https://github.com/kubernetes-csi/csi-driver-smb/blob/master/test/utils/testutil/testutil.go#L62

Can we reuse that?

Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:344]: In windows the path is supposed to be within the context of the value that was passed as this an argument to `--kubelet-pod-path` of csi-proxy binary The default is `C:\\var\\lib\\kubelet\\`. Since this is a new function, that required within the scope of the file `nodeserver_test.go`, I created a new function.

File: pkg/smb/nodeserver_test.go (hunk 7 above)
Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:226]: Thanks will update the PR accordingly.

Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:226]: I think using we might have to make some modification to the existing struct: 
```
type TestError struct {
	WindowsError error
	DefaultError error
}
```
In this struct, we are taking the DefaultError when the WindowsError is `nil`. But there are cases where we are explicitly asking to check with nil. In such cases its bit difficult to check for errors. So I propose a small modification to the struct to have a fields like:
```
type TestError struct {
        expectedError error
	WindowsError error
	DefaultError error
}
So that explictly checking for windows error even if its nil is easier.

what are your thoughts on this?

Comment (mayankshah1607) [reply pkg/smb/nodeserver_test.go:226]: @boddumanohar We're actually following the current approach in the azurefile and azuredisk repos for unit testing - https://github.com/kubernetes-sigs/azurefile-csi-driver , https://github.com/kubernetes-sigs/azuredisk-csi-driver

We didn't really have to make changes to the assertion logic there. It seems like you're suggesting this change because of the unexpected error given by `TestNodeUnpublishVolume` (`[Error] Unmount error mocked by IsLikelyNotMountPoint`) and `TestNodeUnstageVolume` (`[Error] CleanupMountPoint error mocked by IsLikelyNotMountPoint`). IMO, because these are unexpected errors from the CSI proxy, its best if we temporarily skip only these particular test cases on Windows until we figure out why CSI proxy is giving that issue, rather that changing the assertion logic to incorporate this unexpected failures.

I've opened a PR here with the above thoughts, and keeping the same approach as azuredisk and azurefile repos - https://github.com/kubernetes-csi/csi-driver-smb/pull/123

@boddumanohar @andyzhangx Please let me know your thoughts, and we can make more changes accordingly. Thanks! :)

Comment (andyzhangx) [reply pkg/smb/nodeserver_test.go:226]: I think that's case by case
 - `error mocked by IsLikelyNotMountPoint` error is likely caused by no fake of `IsLikelyNotMountPoint` on Linux, so need to fix it on Linux, that may fix the issue?
- `[Error] Mount error mocked by Mount`
  - it's caused by different mount behavior between linux and windows, for this case, let's only run on Linux

In general, I think we should use https://github.com/kubernetes-csi/csi-driver-smb/blob/master/test/utils/testutil/testutil.go#L28, it's more clear.

And if there is one case that only Linux would return error(Windows return nil), we could make it separate, e.g. add `OnlyRunOnPlatform` in test case struct, that would make logic more clear: positive and negative cases should be separated

Comment (andyzhangx) [reply pkg/smb/nodeserver_test.go:226]: point is `TestError` should really contains error, if it's nil, then it's not error. if there is one case that only Linux would return error(Windows return nil), we could make it separate, e.g. add `OnlyRunOnPlatform` in test case struct, that would make logic more clear: positive and negative cases should be separated

Comment (boddumanohar) [reply pkg/smb/nodeserver_test.go:226]: Cool. Let me try this out.

**Comment by boddumanohar** [comment]
Since most of the comments in this PR have been addressed by @mayankshah1607  in his PR#123. I've removed the changes related to tests and kept the original basis of this PR i.e to start CSI Proxy before running windows tests.

**Comment by mayankshah1607** [comment]
Thanks @boddumanohar !
I will rebase my PR once this one is merged. :)

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve
/ok-to-test
